### PR TITLE
Proposal for dealing with requires arguments at parsetime

### PR DESCRIPTION
**Pull request by lukedirtwalker**
We use exceptions as discussed. To specify rules we have various convenience constructs.

For now it is just used in AstQuery, but the goal is to apply it on all queries.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42526043%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42526347%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42527048%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42527646%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42531984%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42532475%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42533830%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42534362%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23issuecomment-149883213%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23issuecomment-149901871%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23issuecomment-149883213%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22So%20I%20fixed%20the%20things%20you%20mentioned%20and%20adapted%20some%20things.%20It%20should%20now%20be%20ready%20to%20merge.%22%2C%20%22created_at%22%3A%20%222015-10-21T12%3A42%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks.%22%2C%20%22created_at%22%3A%20%222015-10-21T13%3A49%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/parsing/ArgumentRule.h%2060%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42533830%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Use%20plain-old%20functions%20here.%22%2C%20%22created_at%22%3A%20%222015-10-20T18%3A23%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/ArgumentRule.h%3AL1-78%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/parsing/ArgumentParser.cpp%2026%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42526043%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22We%20should%20probably%20call%20this%20something%20a%20bit%20more%20descriptive%2C%20e.g.%20%60QueryParsingException%60.%22%2C%20%22created_at%22%3A%20%222015-10-20T17%3A26%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/ArgumentParser.cpp%3AL42-69%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/queries/AstQuery.cpp%2041%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42532475%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22ideally%20you%20should%20not%20have%20to%20provide%20a%20string%2C%20but%20it%20should%20be%20automatically%20deduced.%22%2C%20%22created_at%22%3A%20%222015-10-20T18%3A13%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL73-96%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/parsing/ArgumentParser.cpp%2045%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42526347%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20method%20looks%20a%20bit%20unnecessary.%20Can%27t%20%60rule.check%60%20throw%20itself%3F%20Why%20do%20we%20need%20this%20wrapper%3F%22%2C%20%22created_at%22%3A%20%222015-10-20T17%3A28%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/ArgumentParser.cpp%3AL42-69%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/queries/AstQuery.cpp%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42534362%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20don%27t%20need%20%60ArgumentParser%60%20to%20be%20a%20pointer%20type%20at%20all.%20This%20is%20only%20needed%20for%20polymorphism.%20Use%20a%20plain-old%20type.%22%2C%20%22created_at%22%3A%20%222015-10-20T18%3A27%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL53-71%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/parsing/ArgumentRule.cpp%2038%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42527048%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%60bind%60%20is%20evli%20%28don%27t%20know%20exactly%20why%29%2C%20and%20will%20be%20deprecated.%20Use%20lambdas%20instead.%22%2C%20%22created_at%22%3A%20%222015-10-20T17%3A33%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Maybe%20I%27m%20wrong%20about%20this%2C%20couldn%27t%20find%20a%20conclusive%20source%20for%20this%20info.%22%2C%20%22created_at%22%3A%20%222015-10-20T17%3A37%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/ArgumentRule.cpp%3AL1-74%22%7D%2C%20%22Pull%20ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87%20InformationScripting/src/parsing/ArgumentRule.cpp%2065%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/177%23discussion_r42531984%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22use%20%60std%3A%3Acount_if%60%20and%20make%20this%20look%20identical%20to%20the%20ones%20above.%22%2C%20%22created_at%22%3A%20%222015-10-20T18%3A10%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/ArgumentRule.cpp%3AL1-74%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/parsing/ArgumentParser.cpp 26'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42526043'>File: InformationScripting/src/parsing/ArgumentParser.cpp:L42-69</a></b>
- <a href='https://github.com/dimitar-asenov'></a> We should probably call this something a bit more descriptive, e.g. `QueryParsingException`.
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/parsing/ArgumentParser.cpp 45'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42526347'>File: InformationScripting/src/parsing/ArgumentParser.cpp:L42-69</a></b>
- <a href='https://github.com/dimitar-asenov'></a> This method looks a bit unnecessary. Can't `rule.check` throw itself? Why do we need this wrapper?
- [ ] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/parsing/ArgumentRule.cpp 38'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42527048'>File: InformationScripting/src/parsing/ArgumentRule.cpp:L1-74</a></b>
- <a href='https://github.com/dimitar-asenov'></a> `bind` is evli (don't know exactly why), and will be deprecated. Use lambdas instead.
- <a href='https://github.com/dimitar-asenov'></a> Maybe I'm wrong about this, couldn't find a conclusive source for this info.
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/parsing/ArgumentRule.cpp 65'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42531984'>File: InformationScripting/src/parsing/ArgumentRule.cpp:L1-74</a></b>
- <a href='https://github.com/dimitar-asenov'></a> use `std::count_if` and make this look identical to the ones above.
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/queries/AstQuery.cpp 41'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42532475'>File: InformationScripting/src/queries/AstQuery.cpp:L73-96</a></b>
- <a href='https://github.com/dimitar-asenov'></a> ideally you should not have to provide a string, but it should be automatically deduced.
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/parsing/ArgumentRule.h 60'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42533830'>File: InformationScripting/src/parsing/ArgumentRule.h:L1-78</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Use plain-old functions here.
- [x] <a href='#crh-comment-Pull ff9b37eb1a1b0aad8e2b82a6174b2e2dec283c87 InformationScripting/src/queries/AstQuery.cpp 8'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#discussion_r42534362'>File: InformationScripting/src/queries/AstQuery.cpp:L53-71</a></b>
- <a href='https://github.com/dimitar-asenov'></a> You don't need `ArgumentParser` to be a pointer type at all. This is only needed for polymorphism. Use a plain-old type.
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/177#issuecomment-149883213'>General Comment</a></b>
- <a href='https://github.com/lukedirtwalker'></a> So I fixed the things you mentioned and adapted some things. It should now be ready to merge.
- <a href='https://github.com/dimitar-asenov'></a> Thanks.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/177?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/177?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/177'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dimitar-asenov** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/177'>Refresh</a>.<a href='#crh-update-none'></a>

File: InformationScripting/src/parsing/ArgumentParser.cpp [hunk 1]
```diff
@@ -41,21 +42,27 @@ ScopedArgumentQuery::ScopedArgumentQuery(Model::Node* target, std::initializer_l
 	argParser_->setSingleDashWordOptionMode(QCommandLineParser::ParseAsLongOptions);
 
 	if (!argParser_->parse(args))
-		qWarning() << args[0] << "parse failure"; // TODO warn user
+		throw ParsingException("General parsing error");
```
Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentParser.cpp]: We should probably call this something a bit more descriptive, e.g. `QueryParsingException`.

File: InformationScripting/src/parsing/ArgumentParser.cpp [hunk 2]
```diff
@@ -41,21 +42,27 @@ ScopedArgumentQuery::ScopedArgumentQuery(Model::Node* target, std::initializer_l
...
 	return argParser_->value(argName);
 }
 
-bool ScopedArgumentQuery::isArgumentSet(const QString& argName) const
+bool ArgumentParser::isArgumentSet(const QString& argName) const
 {
 	return argParser_->isSet(argName);
 }
 
+void InformationScripting::ArgumentParser::checkRule(const Arguments::ArgumentRule& rule) const
```
Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentParser.cpp]: This method looks a bit unnecessary. Can't `rule.check` throw itself? Why do we need this wrapper?

File: InformationScripting/src/parsing/ArgumentRule.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include "ArgumentParser.h"
+
+namespace InformationScripting {
+
+namespace Arguments {
+
+ArgumentRule::ArgumentRule(ArgumentCheck checkFunction, QString message, std::vector<ArgumentValue> expectedArguments)
+	: violationMessage{message}
+{
+	check = std::bind(checkFunction, std::placeholders::_1, expectedArguments);
```
Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentRule.cpp]: `bind` is evli (don't know exactly why), and will be deprecated. Use lambdas instead.

Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentRule.cpp]: Maybe I'm wrong about this, couldn't find a conclusive source for this info.

File: InformationScripting/src/parsing/ArgumentRule.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+bool RequireOneOf::operator()(const ArgumentParser& parser, const std::vector<ArgumentValue>& values) const
+{
+	return std::any_of(values.begin(), values.end(), ArgumentValueCheck{parser});
+}
+
+bool AtMostOneOf::operator()(const ArgumentParser& parser, const std::vector<ArgumentValue>& values) const
+{
+	auto p = ArgumentValueCheck{parser};
```
Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentRule.cpp]: use `std::count_if` and make this look identical to the ones above.

File: InformationScripting/src/queries/AstQuery.cpp [hunk 5]
```diff
@@ -69,16 +73,23 @@ Optional<TupleSet> AstQuery::executeLinear(TupleSet input)
...
+	registerQuery("toClass", &AstQuery::toParentType, {}, "Class");
 	registerQuery("bases", &AstQuery::baseClassesQuery);
 	registerQuery("callgraph", &AstQuery::callGraph);
-	registerQuery("ast", &AstQuery::genericQuery);
-	registerQuery("toParent", &AstQuery::toParentType);
-	registerQuery("uses", &AstQuery::usesQuery);
-	registerQuery("type", &AstQuery::typeFilter);
-	registerQuery("attribute", &AstQuery::attribute);
+	registerQuery("ast", &AstQuery::genericQuery,
+		{{Arguments::RequireOneOf(), "ast requires either name or type argument",
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/AstQuery.cpp]: ideally you should not have to provide a string, but it should be automatically deduced.

File: InformationScripting/src/parsing/ArgumentRule.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+// Convenience Rules:
+class ArgumentValueCheck {
+	public:
+		ArgumentValueCheck(const ArgumentParser& parser);
+		bool operator() (const ArgumentValue& value) const;
+	private:
+		const ArgumentParser& parser_;
+};
+
+struct RequireAll
```
Comment (dimitar-asenov) [inline InformationScripting/src/parsing/ArgumentRule.h]: Use plain-old functions here.

File: InformationScripting/src/queries/AstQuery.cpp [hunk 7]
```diff
@@ -53,14 +53,18 @@ const QStringList AstQuery::NAME_ARGUMENT_NAMES{"n", "name"};
 const QStringList AstQuery::ADD_AS_NAMES{"a", "addAs"};
 const QStringList AstQuery::ATTRIBUTE_NAME_NAMES{"at", "attribute"};
 
-AstQuery::AstQuery(ExecuteFunction exec, Model::Node* target, QStringList args)
-	: ScopedArgumentQuery{target, {
+AstQuery::AstQuery(ExecuteFunction exec, Model::Node* target, QStringList args,
+						 std::vector<Arguments::ArgumentRule> argumentRules)
+	: target_{target}, arguments_{new ArgumentParser({
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/AstQuery.cpp]: You don't need `ArgumentParser` to be a pointer type at all. This is only needed for polymorphism. Use a plain-old type.

**Comment by lukedirtwalker** [comment]
So I fixed the things you mentioned and adapted some things. It should now be ready to merge.

**Comment by dimitar-asenov** [comment]
Thanks.

**Pull request merged**